### Name : Ibadullah Hayat
### Reg no : B23F0001AI010
### Section: F23 AI-Green

## Lab 02: Backpropagation and Its Implementation from Scratch

### Objective
Manually compute one full step of backpropagation for a small neural network to understand how gradients are calculated and weights are updated without using deep learning libraries.

#### Problem Setup
We are given:

Inputs:
x1 = 0.35,
x2 = 0.7

Target output:
y(target) = 0.5

Learning rate:
η=1

Activation function: Sigmoid
σ(z)= 1/1+e^-z

Network architecture:
Input → 2 hidden neurons → 1 output neuron

In [ ]:
import numpy as np

def sigmoid(z):
    """Sigmoid activation function."""
    return 1 / (1 + np.exp(-z))

def sigmoid_derivative(y):
    """Derivative of sigmoid given output y."""
    return y * (1 - y)

Task 1: Forward Pass Analysis

Compute prediction step-by-step

In [ ]:
# Inputs and target
x1, x2 = 0.35, 0.7
y_target = 0.5

# Initial weights
w11, w21 = 0.2, 0.2  # to h1
w12, w22 = 0.3, 0.3  # to h2
w13, w23 = 0.3, 0.9  # to output

# Hidden layer weighted sums
a1 = w11 * x1 + w21 * x2  # h1 input
a2 = w12 * x1 + w22 * x2  # h2 input

# Hidden layer outputs
y3 = sigmoid(a1)  # h1 output
y4 = sigmoid(a2)  # h2 output

# Output layer
a3 = w13 * y3 + w23 * y4
y5 = sigmoid(a3)

print(f"Weighted sum at h1 (a1): {a1:.4f}")
print(f"Output of h1 (y3):       {y3:.4f}")
print(f"Weighted sum at h2 (a2): {a2:.4f}")
print(f"Output of h2 (y4):       {y4:.4f}")
print(f"Final output (y5):       {y5:.4f}")
print(f"Target:                  {y_target}")
print(f"Error (target - pred):   {y_target - y5:.4f}")

Weighted sum at h1 (a1): 0.2100
Output of h1 (y3):       0.5523
Weighted sum at h2 (a2): 0.3150
Output of h2 (y4):       0.5781
Final output (y5):       0.6651
Target:                  0.5
Error (target - pred):   -0.1651


Interpretation: Prediction (0.6651) > target (0.5) → overestimates risk.

Task 2: Error Calculation

In [ ]:
error = y_target - y5
print(f"Error = y_target - y5 = {error:.4f}")
if error < 0:
    print("→ Prediction is TOO HIGH.")
else:
    print("→ Prediction is TOO LOW.")

Error = y_target - y5 = -0.1651
→ Prediction is TOO HIGH.



Task 3: Output Neuron Responsibility (δ₅)

In [ ]:
delta5 = sigmoid_derivative(y5) * error
print(f"δ₅ = y5(1 - y5)(error) = {delta5:.6f}")

δ₅ = y5(1 - y5)(error) = -0.036770


Why multiply by derivative?

The derivative tells us how sensitive the output is to changes in input. Near saturation (0 or 1), learning slows - this prevents large updates when the neuron is "confident but wrong."

Task 4: Hidden Neuron Responsibility (δ₃, δ₄)

In [ ]:
delta3 = sigmoid_derivative(y3) * w13 * delta5
delta4 = sigmoid_derivative(y4) * w23 * delta5

print(f"δ₃ = y3(1-y3) * w13 * δ5 = {delta3:.6f}")
print(f"δ₄ = y4(1-y4) * w23 * δ5 = {delta4:.6f}")

δ₃ = y3(1-y3) * w13 * δ5 = -0.002728
δ₄ = y4(1-y4) * w23 * δ5 = -0.008071


Task 5: Weight Updates

In [ ]:
eta = 1.0

# Update hidden → output weights
dw13 = eta * delta5 * y3
dw23 = eta * delta5 * y4
w13_new = w13 - dw13
w23_new = w23 - dw23

# Update input → hidden weights
dw11 = eta * delta3 * x1
dw21 = eta * delta3 * x2
dw12 = eta * delta4 * x1
dw22 = eta * delta4 * x2

w11_new = w11 - dw11
w21_new = w21 - dw21
w12_new = w12 - dw12
w22_new = w22 - dw22

print("=== Updated Weights ===")
print(f"w13: {w13:.3f} → {w13_new:.4f} (Δ = {dw13:.6f})")
print(f"w23: {w23:.3f} → {w23_new:.4f} (Δ = {dw23:.6f})")
print(f"w11: {w11:.3f} → {w11_new:.4f} (Δ = {dw11:.6f})")
print(f"w21: {w21:.3f} → {w21_new:.4f} (Δ = {dw21:.6f})")
print(f"w12: {w12:.3f} → {w12_new:.4f} (Δ = {dw12:.6f})")
print(f"w22: {w22:.3f} → {w22_new:.4f} (Δ = {dw22:.6f})")

=== Updated Weights ===
w13: 0.300 → 0.3203 (Δ = -0.020309)
w23: 0.900 → 0.9213 (Δ = -0.021257)
w11: 0.200 → 0.2010 (Δ = -0.000955)
w21: 0.200 → 0.2019 (Δ = -0.001909)
w12: 0.300 → 0.3028 (Δ = -0.002825)
w22: 0.300 → 0.3056 (Δ = -0.005650)


Forward pass after weight updates

In [ ]:
# Recompute forward pass with new weights
a1_new = w11_new * x1 + w21_new * x2
a2_new = w12_new * x1 + w22_new * x2
y3_new = sigmoid(a1_new)
y4_new = sigmoid(a2_new)
a3_new = w13_new * y3_new + w23_new * y4_new
y5_new = sigmoid(a3_new)

print(f"Old prediction: {y5:.4f}")
print(f"New prediction: {y5_new:.4f}")
print(f"Target:         {y_target}")
print(f"Error reduced:  {abs(y_target - y5) - abs(y_target - y5_new):.4f}")

Old prediction: 0.6651
New prediction: 0.6706
Target:         0.5
Error reduced:  -0.0055


Even after one update, the network moves toward the target—learning is working!

Task 6: Interpretation & Reflection

1. Explain backpropagation as blame assignment.
Backpropagation distributes responsibility for the output error backward through the network. Each neuron’s error term (δ) reflects how much it contributed to the mistake—based on its activation and outgoing weights.

2. Effect of learning rate
Too large: Overshooting → divergence or oscillation.
Too small: Extremely slow convergence; may get stuck in local minima.
3. Why “backward” propagation?
Because we start from the output error and propagate gradients backward through layers using the chain rule-opposite to data flow.


### Summary
We manually implemented one full training step of backpropagation.

Verified that weight updates reduce prediction error.

Gained intuition about gradient flow, activation derivatives, and learning dynamics.

This foundational understanding is what makes modern deep learning libraries (PyTorch, TensorFlow) powerful - they automate this process, but you now know what’s under the hood.